FLT/DivisionAlgebra/Finiteness.lean

```diff
@@ -212,19 +309,175 @@ abbrev Dfx := (D ⊗[K] (FiniteAdeleRing (𝓞 K) K))ˣ
 noncomputable abbrev incl₁ : Dˣ →* Dfx K D :=
   Units.map Algebra.TensorProduct.includeLeftRingHom.toMonoidHom
 
+variable [MeasurableSpace (D ⊗[K] NumberField.AdeleRing (𝓞 K) K)]
+  [BorelSpace (D ⊗[K] NumberField.AdeleRing (𝓞 K) K)]
+
+def iso₁ : (D ⊗[K] NumberField.AdeleRing (𝓞 K) K)ˣ ≃*
+    Prod (D ⊗[K] NumberField.InfiniteAdeleRing K)ˣ (Dfx K D) := by
+  simp_rw [NumberField.AdeleRing, Dfx]
+  have h1 : D ⊗[K] (NumberField.InfiniteAdeleRing K × FiniteAdeleRing (𝓞 K) K) ≃
+      (D ⊗[K] (NumberField.InfiniteAdeleRing K)) × (D ⊗[K] FiniteAdeleRing (𝓞 K) K) := by
+    -- DirectSum.addEquivProd
+    -- TensorProduct.directSumRight
+
+    sorry
+  have := Units.mapEquiv h1
+  apply MulEquiv.prodUnits at this
+  sorry
+
+def rest₁ : ringHaarChar_ker D_𝔸 → Dfx K D :=
+  fun a => (iso₁ K D) a.val |>.2
```

Comment:
It's not clear to me that making definition after definition is a wise idea in general. Every definition clogs the system up. Sure abbreviate things if you're doing paper calculations, but every time you make a definition you should be making API for the definition, and here you make no API and immediately go on to make another definition. Definitions should be few and far between.